### PR TITLE
Ethereum compatible merkle trie hasher

### PR DESCRIPTION
**Pull request by JcBernack**
- add merkle trie hasher to libevm
- add interop code
- refactor evm interop tests to multiple units

### CODE REVIEW

File: evm/src/test/java/com/horizen/evm/TrieHasherTest.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        assertArrayNotEquals("should not give empty root hash", emptyRootHash, hashA);
+        assertArrayEquals("should return same root hash for same input", hashA, hashA2);
+        assertArrayNotEquals("should return different root hash for different input", hashA, hashB);
+        assertArrayNotEquals("should return different root hash for different input", hashB, hashC);
+        assertArrayNotEquals("should return different root hash for different input", hashC, hashD);
+        assertArrayNotEquals("should return different root hash for different input", hashD, hashE);
+        assertArrayNotEquals("should return different root hash for different input", hashE, hashF);
+        assertArrayNotEquals("should return different root hash for different input", hashF, hashG);
+
+        // TODO: generate some receipts, RLP encode them and verify the root hash
```
Comment (i-Alex) [inline evm/src/test/java/com/horizen/evm/TrieHasherTest.java:47]: yes, would be nice to have a regression test for receipts, where to check the result against the one retrieved in `func TestHashRoot(t *testing.T)`

File: libevm/lib/service_hash.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+// HashRoot returns the root hash of the merkle trie after adding all the given values using their index as the key
+// heavily based on types.Derive(), see https://github.com/ethereum/go-ethereum/blob/d12b1a91cd9423f83bf77dbe363164797549ff15/core/types/hashing.go#L87
+func (s *Service) HashRoot(params HashParams) (error, common.Hash) {
+	hasher := trie.NewStackTrie(nil)
+	count := len(params.Values)
+	// StackTrie requires values to be inserted in increasing hash order, which is not the
+	// order that `list` provides hashes in. This insertion sequence ensures that the
+	// order is correct.
+	var indexBuf []byte
+	for i := 1; i < count && i <= 0x7f; i++ {
```
[the remark this answers is not in the document]
Comment (JcBernack) [reply libevm/lib/service_hash.go:22]: Absolutely. Took me a while to figure this one out:
This is build on the knowledge of how the binary representation of the index will be after RLP encoding.

The index will be represented as a byte array in a big-endian style encoding, but any leading zero-values are removed. The resulting byte array is then encoded like this:
- index 1 to 127 will result in a byte array with one value and encoded as-is, i.e. 0x01 to 0x7F
- index value 0 will result in an empty byte array which is encoded as `length of the array + 0x80`, i.e. 0x80
- index values 128 and larger will result in a byte array with 1 or more elements which are prefixed by `length of the array + 0x80`, e.g. 0x8180 (128) or 0x82012C (300)

This way the key-value pairs are added to the trie in the order of increasing indices.

[the remark this answers is not in the document]
Comment (JcBernack) [reply libevm/lib/service_hash.go:22]: Source: Ethereum Yellow Paper Appendix B: RLP

[the remark this answers is not in the document]
Comment (i-Alex) [reply libevm/lib/service_hash.go:22]: quite tricky, but now is clear.
Thanks!

**Pull request merged**